tests/scenario/test_charm.py

```diff
@@ -0,0 +1,137 @@
+import re
+from contextlib import ExitStack, contextmanager
+from unittest.mock import patch, PropertyMock
+
+import pytest
+from ops import testing
+
+from charm import GrafanaCharm
+
+
+@contextmanager
+def grafana_ready(ready: bool):
+    with patch("grafana_client.Grafana.is_ready", PropertyMock(return_value=ready)):
+        yield
+
+
+@contextmanager
+def password_changed(changed: bool):
+    with patch("grafana_client.Grafana.password_has_been_changed", return_value=changed):
+        yield
+
+
+@pytest.fixture(autouse=True)
+def patch_all():
+    with ExitStack() as stack:
+        for p in [
+            patch("lightkube.core.client.GenericSyncClient"),
+            patch("socket.getfqdn", new=lambda *args: "grafana-k8s-0.testmodel.svc.cluster.local"),
+            patch("socket.gethostbyname", new=lambda *args: "1.2.3.4"),
+            patch.multiple(
+                "charm.KubernetesComputeResourcesPatch",
+                _namespace="test-namespace",
+                _patch=lambda *a, **kw: True,
+                is_ready=lambda *a, **kw: True,
+            ),
+            patch.object(GrafanaCharm, "grafana_version", "0.1.0"),
+        ]:
+            stack.enter_context(p)
+
+        yield
+
+
+@pytest.fixture
+def ctx():
+    return testing.Context(GrafanaCharm)
+
+
+def check_valid_password(s: str):
+    assert re.match(r"[A-Za-z0-9]{12}", s)
+
+
+def test_can_get_password(ctx):
+    # GIVEN a grafana leader unit
+    state = testing.State(leader=True)
+
+    # WHEN we receive any hook
+    with ctx(ctx.on.update_status(), state) as mgr:
+        # THEN the .admin_password attribute returns a valid password
+        pwd = mgr.charm.admin_password
+        check_valid_password(pwd)
```

Comment:
Per interface, `def admin_password(self) -> Optional[str]:` and test scope, how about:
```suggestion
        # THEN the .admin_password attribute returns a string with length > 0
        pwd = mgr.charm.admin_password
        assert isinstance(pwd, str)
        assert len(pwd) > 0
```